### PR TITLE
Replace RelationSource with Analyzed-/QueriedRelation

### PR DESCRIPTION
**Pull request by mfussenegger**
In order to do a "query push-down" we used a custom data structure to
represent a query with a JOIN.

For example

    select t1.x, t2.x from t1, t2 where t1.x = 10

Was represented as:

    select t1.x, t2.x from t1, t2

    t1 querySpec:
        outputs: [x]
        where: x = 10

    t2 querySpec:
        outputs: [x]

This commit changes the internal representation so that the sub-query
data structures are used. The above example now looks like this:

    select t1.x, t2.x from
        (select x from t1 where x = 1) t1,
        (select x from t2) t2

This should make it easier to support joined virtual tables since the
internal representation is already the same.

### CODE REVIEW

**Comment by mfussenegger** [comment]
@matriv can you do an initial review of the last commit? The other commits are in a separate PR

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java [hunk 1]
```diff
@@ -23,46 +23,95 @@
...
+            mss.sources.values(),
             mss.joinPairs
         );
         splitter.process();
 
-        for (RelationSource source : mss.sources.values()) {
-            QuerySpec spec = splitter.getSpec(source.relation());
-            source.querySpec(spec);
+        Function<Field, Field> convertFieldToPointToNewRelations = Function.identity();
+        Function<? super Symbol, ? extends Symbol> convertFieldInSymbolsToNewRelations = Function.identity();
```
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: *ToPointTo*NewRelations

File: sql/src/main/java/io/crate/analyze/Relations.java [hunk 2]
```diff
@@ -35,4 +38,25 @@
     static Collection<? extends Path> namesFromOutputs(List<Symbol> outputs) {
         return Lists.transform(outputs, Symbols::pathFromSymbol);
     }
+
+    static QueriedRelation upgrade(Functions functions,
```
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/Relations.java]: Add short comments

File: sql/src/main/java/io/crate/analyze/OrderBy.java [hunk 3]
```diff
@@ -79,7 +79,7 @@ public OrderBy subset(Collection<Integer> positions) {
         boolean[] reverseFlags = new boolean[positions.size()];
         int pos = 0;
         for (Integer i : positions) {
-            orderBySymbols.add(this.orderBySymbols.get(i));
+            orderBySymbols.add(Symbols.DEEP_COPY.apply(this.orderBySymbols.get(i)));
```
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/OrderBy.java]: Why is this needed here?

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java (hunk 1 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: do you think that's necessary? I think the names are already long enough as it is and thought it'd be okay to omit the "pointTo"

File: sql/src/main/java/io/crate/analyze/OrderBy.java (hunk 3 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/OrderBy.java]: Otherwise the symbols are mutated somewhere (I think due to normalization) and then the InputColumns cannot be created anymore because of a field/reference miss-match.

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java (hunk 1 above)
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: yeah ok :-)

File: sql/src/main/java/io/crate/analyze/Rewriter.java [hunk 4]
```diff
@@ -142,61 +139,68 @@ private static boolean couldMatchWithNullValues(EvaluatingNormalizer normalizer,
...
             } else {
                 joinPair.joinType(JoinType.RIGHT);
             }
             multiSourceQuerySpec.where(new WhereClause(AndOperator.join(splitQueries.values())));
         }
         for (Field fieldToRemove : collectFieldsToRemoveFromOutputs.fieldsToNotCollect()) {
-            outerSpec.outputs().remove(fieldToRemove);
             multiSourceQuerySpec.outputs().remove(fieldToRemove);
+            AnalyzedRelation relation = fieldToRemove.relation();
+            if (relation instanceof QueriedRelation) {
```
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/Rewriter.java]: Remove if the check not needed

File: sql/src/main/java/io/crate/planner/fetch/MultiSourceFetchPushDown.java [hunk 5]
```diff
@@ -57,46 +65,54 @@
...
+                for (int i = 0; i < relation.querySpec().outputs().size(); i++) {
+                    Symbol output = relation.querySpec().outputs().get(i);
                     if (!canBeFetched.contains(output)) {
                         qtOutputs.add(output);
-                        mssOutputs.add(output);
+                        mssOutputs.add(relation.fields().get(i));
                     }
                 }
-                for (Field field : canBeFetched) {
+                for (Tuple<Field, Reference> parentAndChild : canBeFetched.parentAndChildren()) {
```
Comment (matriv) [inline sql/src/main/java/io/crate/planner/fetch/MultiSourceFetchPushDown.java]: Please some comment to explain what's happening here (and block after the for loop). thx!

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java [hunk 6]
```diff
@@ -23,56 +23,107 @@
...
+     *     select t1.x, t2.y from t1, t2 where t1.x = 10
+     *
+     *     becomes
+     *
+     *     select t1.x, t2.y from
+     *          (select x from t1 where x = 10) t1,
+     *          (select y from t2)
+     * </pre>
+     */
+    public static MultiSourceSelect createWithPushDown(Functions functions,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: Think we could consider having a new MultiSourceSelect class that has `sources <QualifiedName, QueriedRelation>` to avoid all the casts that became otherwise necessary. Thoughts?

File: sql/src/main/java/io/crate/planner/consumer/QueryAndFetchConsumer.java [hunk 7]
```diff
@@ -198,11 +198,13 @@ private static boolean tryApplyOrderAndLimitPreFetch(QuerySpec qs,
                                                          Limits limits) {
         OrderBy orderBy = qs.orderBy().orElse(null);
         if (orderBy == null) {
-            TopNProjection topN = new TopNProjection(
-                limits.finalLimit(),
-                limits.offset(),
-                InputColumn.fromSymbols(fetchDescription.preFetchOutputs()));
-            plan.addProjection(topN, null, null, null);
+            if (limits.hasLimit()) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/planner/consumer/QueryAndFetchConsumer.java]: this is not really relevant for this PR - I noticed this when I tried to remove the joins on subQuery restriction - should I revert it for now?

File: sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java [hunk 8]
```diff
@@ -110,6 +111,18 @@ private void processOutputs() {
             });
         }
 
+        Optional<List<Symbol>> groupBy = querySpec.groupBy();
+        if (groupBy.isPresent()) {
+            FieldsVisitor.visitFields(groupBy.get(), addFieldToMap);
+        }
+        Optional<HavingClause> having = querySpec.having();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java]: I could probably do this in a separate PR upfront and create unit tests - I think this wasn't required so far because of how the fetch-rewrite worked (It didn't have strict relation boundaries as we have now that the MSS sources are a QueriedRelation) 

(This is about adding GROUP-KEYS / HAVING expressions which aren't in the outputs to the output of the newly created sub-relations)

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java (hunk 6 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: Another option would be to initially create the MSS already with QueriedRelations as sources. In the form of

    select t1.x, t2.x from (select * from t1) t1, (select * from t2) t2

the pushdown would then just "narrow" / prune the select-list and push-down other clauses.

File: sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java (hunk 8 above)
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java]: Yep I agree to make it in separate PR

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java (hunk 6 above)
Comment (matriv) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: If not very complex, I'd like that, as you've agreed those (QueriedRelation) casts are ugly...

File: sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java (hunk 8 above)
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java]: +1

File: sql/src/main/java/io/crate/planner/consumer/QueryAndFetchConsumer.java (hunk 7 above)
Comment (dobe) [inline sql/src/main/java/io/crate/planner/consumer/QueryAndFetchConsumer.java]: seperate PR before?

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java (hunk 6 above)
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: if this is no problem when you mutate the outputs of the queriedrelations, then i would do it

File: sql/src/main/java/io/crate/analyze/Rewriter.java [hunk 9]
```diff
@@ -142,61 +139,65 @@ private static boolean couldMatchWithNullValues(EvaluatingNormalizer normalizer,
         };
     }
 
-    private static Map<QualifiedName, QuerySpec> groupOuterRelationQSByName(Map.Entry<QualifiedName, RelationSource> left,
-                                                                            Map.Entry<QualifiedName, RelationSource> right,
+    private static Map<QualifiedName, QueriedRelation> groupOuterRelationQSByName(Map.Entry<QualifiedName, AnalyzedRelation> left,
+                                                                            Map.Entry<QualifiedName, AnalyzedRelation> right,
```
Comment (seut) [inline sql/src/main/java/io/crate/analyze/Rewriter.java]: pls fix format

File: sql/src/main/java/io/crate/analyze/MultiSourceSelect.java (hunk 6 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/MultiSourceSelect.java]: FTR after short discussion:

- Having two separate MultiSourceSelect classes (one with AnalyzedRelation and one with QueriedRelation sources) is probably not worth it
- We need an AnalyzedRelation early on: Before the query-splitter we don't know the outputs that are needed and `fields()` of a TableRelation only contains top-level columns, so accessing subscripts would become impossible without introducing some kind of RowType. That's out-of-scope for now.

So we'll keep the casts for now.

File: sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java (hunk 8 above)
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java]: Done via https://github.com/crate/crate/pull/5538

File: sql/src/main/java/io/crate/analyze/Relations.java [hunk 10]
```diff
@@ -35,4 +38,36 @@
...
+     * </pre>
+     *
+     * If the result is a QueriedTable it is also normalized.
+     */
+    static QueriedRelation applyQSToRelation(Functions functions,
+                                             TransactionContext transactionContext,
+                                             AnalyzedRelation relation,
+                                             QuerySpec querySpec) {
+        QueriedRelation newRelation;
+        if (relation instanceof DocTableRelation) {
```
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/Relations.java]: why not use a visitor?

Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/Relations.java]: using a Visitor would require a Context class because there are 3 arguments that are required in addition to the relation (functions, transactionContext, querySpec). `instanceof` should be cheaper.

Comment (dobe) [inline sql/src/main/java/io/crate/analyze/Relations.java]: hm, an anynymous visitor override would also work, which is of course also an object, however i leave it up to you. looks like a micro optimization
